dev-tools/releng/src/main.rs

```diff
@@ -0,0 +1,669 @@
+    std::env::set_current_dir(&*WORKSPACE_DIR)
+        .context("failed to change working directory to workspace root")?;
+
+    // Unset `$CARGO`, `$CARGO_MANIFEST_DIR`, and `$RUSTUP_TOOLCHAIN` (all
+    // set by cargo or its rustup proxy), which will interfere with various
+    // tools we're about to run. (This needs to come _after_ we read from
+    // `WORKSPACE_DIR` as it relies on `$CARGO_MANIFEST_DIR`.)
+    for var in ["CARGO", "CARGO_MANIFEST_DIR", "RUSTUP_TOOLCHAIN"] {
+        debug!(logger, "unsetting ${}", var);
+        std::env::remove_var(var);
+    }
+
+    // Now that we're done mucking about with our environment (something that's
+    // not necessarily safe in multi-threaded programs), create a Tokio runtime
+    // and call `do_run`.
+    do_run(logger, args)
+}
+
+#[tokio::main]
+async fn do_run(logger: Logger, args: Args) -> Result<()> {
+    let permits = Arc::new(Semaphore::new(num_cpus::get()));
+
+    let commit = Command::new("git")
+        .args(["rev-parse", "HEAD"])
+        .ensure_stdout(&logger)
+        .await?
+        .trim()
+        .to_owned();
+
+    let mut version = BASE_VERSION.clone();
+    // Differentiate between CI and local builds. We use `0.word` as the
+    // prerelease field because it comes before `alpha`.
+    version.pre =
+        if std::env::var_os("CI").is_some() { "0.ci" } else { "0.local" }
+            .parse()?;
+    // Set the build metadata to the current commit hash.
+    let mut build = String::with_capacity(14);
+    build.push_str("git");
+    build.extend(commit.chars().take(11));
+    version.build = build.parse()?;
+    info!(logger, "version: {}", version);
+
+    let manifest = Arc::new(omicron_zone_package::config::parse_manifest(
+        &fs::read_to_string(WORKSPACE_DIR.join("package-manifest.toml"))
+            .await?,
+    )?);
+    let opte_version =
+        fs::read_to_string(WORKSPACE_DIR.join("tools/opte_version")).await?;
+
+    let client = reqwest::ClientBuilder::new()
+        .connect_timeout(Duration::from_secs(15))
+        .timeout(Duration::from_secs(15))
+        .build()
+        .context("failed to build reqwest client")?;
+
+    // PREFLIGHT ==============================================================
+    let mut preflight_ok = true;
+
+    for package in HOST_IMAGE_PACKAGES
+        .into_iter()
+        .chain(RECOVERY_IMAGE_PACKAGES)
+        .map(|(package, _)| package)
+        .chain(TUF_PACKAGES)
+    {
+        if !manifest.packages.contains_key(package) {
+            error!(
+                logger,
+                "package {} to be installed in the OS image \
+                is not listed in the package manifest",
+                package
+            );
+            preflight_ok = false;
+        }
+    }
+
+    // Ensure the Helios checkout exists
+    if args.helios_dir.exists() {
+        if !args.ignore_helios_origin {
+            // check that our helios clone is up to date
+            Command::new("git")
```

Comment:
Should generally consult the `GIT` env var first, then use `git` if that isn't found (applies to rest of this file as well)